### PR TITLE
Feat/dream_emotion

### PR DESCRIPTION
**Pull request by asleepann**
Made dream_emotion dist with the following annotators: bot-emotion-classifier (determines emotion and mood of the bot) and emotional-bot-response (change hypotheses text according to emotion and mood of the bot). Response selector in dream_emotion selects emotional text of the hypothesis.

### CODE REVIEW

File: annotators/emotional_bot_response/server.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+sentry_sdk.init(dsn=os.getenv("SENTRY_DSN"), integrations=[FlaskIntegration()])
+
+logging.basicConfig(format="%(asctime)s - %(name)s - %(levelname)s - %(message)s", level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+app = Flask(__name__)
+health = HealthCheck(app, "/healthcheck")
+logging.getLogger("werkzeug").setLevel("WARNING")
+
+GENERATIVE_SERVICE_URL = getenv("GENERATIVE_SERVICE_URL", "http://openai-api-chatgpt:8145/respond")
```
Comment (dilyararimovna) [inline annotators/emotional_bot_response/server.py:25]: why is this an annotator?

**Review by dilyararimovna — CHANGES_REQUESTED** [review]

**Review by dilyararimovna — CHANGES_REQUESTED** [review]
work with bantches

File: assistant_dists/dream_emotion/dev.yml [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+      - "./common:/src/common"
+    ports:
+      - 8145:8145
+  dff-dream-persona-chatgpt-prompted-skill:
+    volumes:
+      - "./skills/dff_template_prompted_skill:/src"
+      - "./common:/src/common"
+    ports:
+      - 8137:8137
+  dff-google-api-skill:
```
[the remark this answers is not in the document]
Comment (asleepann) [reply assistant_dists/dream_emotion/dev.yml:126]: этот скилл и не работает, но всё работает без него. я не стала убирать, потому что он есть в основном дистрибутиве дрима

File: annotators/emotional_bot_response/server.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+GENERATIVE_TIMEOUT = int(getenv("GENERATIVE_TIMEOUT", 30))
+N_UTTERANCES_CONTEXT = int(getenv("N_UTTERANCES_CONTEXT", 1))
+
+ENVVARS_TO_SEND = getenv("ENVVARS_TO_SEND", None)
+ENVVARS_TO_SEND = [] if ENVVARS_TO_SEND is None else ENVVARS_TO_SEND.split(",")
+
+available_variables = {f"{var}": getenv(var, None) for var in ENVVARS_TO_SEND}
+
+API_CONFIGS = getenv("API_CONFIGS", None)
+API_CONFIGS = [] if API_CONFIGS is None else API_CONFIGS.split(",")
```
[the remark this answers is not in the document]
Comment (asleepann) [reply annotators/emotional_bot_response/server.py:45]: мне сказали, что лучше переиспользовать существующие сервисы, поэтому обращаюсь к сервису с chatgpt

File: assistant_dists/dream_emotion/telegram.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+services:
+  agent-tg:
+    command: sh -c 'bin/wait && python -m deeppavlov_agent.run agent.channel=telegram agent.telegram_token=$TG_TOKEN agent.pipeline_config=assistant_dists/dream/pipeline_conf.json agent.db_config=assistant_dists/dream/db_conf.json'
```
[the remark this answers is not in the document]
Comment (asleepann) [reply assistant_dists/dream_emotion/telegram.yml]: fixed

File: assistant_dists/dream_emotion/docker-compose.override.yml [hunk 5]
```diff
@@ -0,0 +1,495 @@
+services:
+  agent:
+    command: sh -c 'bin/wait && python -m deeppavlov_agent.run agent.pipeline_config=assistant_dists/dream/pipeline_conf.json'
```
[the remark this answers is not in the document]
Comment (asleepann) [reply assistant_dists/dream_emotion/docker-compose.override.yml]: fixed

File: response_selectors/emotion_ranking_based_response_selector/README.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# Ranking-based Response Selector
```
[the remark this answers is not in the document]
Comment (asleepann) [reply response_selectors/emotion_ranking_based_response_selector/README.md]: fixed

File: state_formatters/dp_formatters.py [hunk 7]
```diff
@@ -1252,3 +1253,49 @@ def dff_command_selector_skill_formatter(dialog: Dict) -> List[Dict]:
...
+    if len(dialog["human_utterances"]) > 1:
+        bot_mood = dialog["human_utterances"][-2]["annotations"].get("bot_emotion_classifier",
+                                                                     {}).get("bot_mood", [0.75, 0.25, 0.44])
+    else:
+        bot_mood = [0.75, 0.25, 0.44]
+
+    return [{"sentences": dialog["human_utterances"][-1]["text"],
+             "user_emotion": user_emotion,
+             "sentiment": sent,
+             "bot_mood": bot_mood}]
```
[the remark this answers is not in the document]
Comment (asleepann) [reply state_formatters/dp_formatters.py]: fixed

File: assistant_dists/dream_emotion/dev.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (asleepann) [reply assistant_dists/dream_emotion/dev.yml]: deleted this skill

File: annotators/emotional_bot_response/server.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+def respond():
+    st_time = time.time()
+
+    sentences = request.json.get("sentences", [])
+    bot_mood_label = request.json.get("bot_mood_label", [])
+    bot_emotion = request.json.get("bot_emotion", [])
+
+    results = []
+    for i in range(len(sentences)):
+        result = rewrite_sentences(sentences[i], bot_emotion[0], bot_mood_label[0])
```
Comment (dilyararimovna) [inline annotators/emotional_bot_response/server.py:174]: you are working with a batches. You have to have all elements of the same length. Even if this means you will have to duplicate some of the elememnts.
So, for you to understand this line should look like this:
```
result = rewrite_sentences(sentences[i], bot_emotion[i], bot_mood_label[i])
```

**Review by dilyararimovna — CHANGES_REQUESTED** [review]

File: state_formatters/dp_formatters.py [hunk 9]
```diff
@@ -1223,32 +1376,127 @@ def prompts_goals_collector_formatter(dialog: Dict) -> List[Dict]:
...
+def user_emotion_bot_mood_formatter(dialog: Dict) -> List[Dict]:
+    # comment out the next two lines to save dialog data to see its contents
+    # with open('test_formatters.json', 'w', encoding='utf-8') as f:
+    #     json.dump(dialog, f, ensure_ascii=False, indent=4)
+    sentences = [dialog["human_utterances"][-1]["text"]]
+
+    user_emotions = (
+        dialog["human_utterances"][-1]["annotations"]
+        .get("combined_classification", {})
+        .get("emotion_classification", {})
```
Comment (dilyararimovna) [inline state_formatters/dp_formatters.py:1435]: why do you do this pre-processing in formatter?
I suggest to move it inside your annotator. And use common.utils.get_emotion, common.utils.get_sentiment functions to have a backward compatibility.

**Review by dilyararimovna — CHANGES_REQUESTED** [review]

File: state_formatters/dp_formatters.py (hunk 9 above)
Comment (asleepann) [reply state_formatters/dp_formatters.py]: moved preprocessing of sentiment and emotion to annotator, used functions from common.utils

File: annotators/emotional_bot_response/server.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+
+
+try:
+    logger.info("emotional-bot-response is starting")
+
+    sentences = ["I will eat pizza."]
+    bot_mood_label = "angry"
+    bot_emotion = "anger"
+    response = rewrite_sentences(sentences, bot_mood_label, bot_emotion)
+    print(response)
```
[the remark this answers is not in the document]
Comment (asleepann) [reply annotators/emotional_bot_response/server.py]: fixed

**Pull request closed without merging**